### PR TITLE
Prevent flushing same tags several times

### LINKED ISSUE #29890: [Issue] Prevent flushing same tags several times
This issue is automatically created based on existing pull request: magento/magento2#29885: Prevent flushing same tags several times

## 
### Preconditions (*)
1. Magento 2.4-develop

### Description (*)
If a product is saved in admin - changed name, price, stock, weight, added category to product. Then after indexing, the same tags are flushed several times.
Here is an example of default Magento behaviour on cleaning cache tags during indexing (different arrays belong to different indexers cleaning the cache):
```
root@c119d6a06fcb:/var/www/m2clean$  bin/magento cron:run --group=index --bootstrap=standaloneProcessStarted=1

Array
(
)


Array
(
)


Array
(
    <0] => cat*p*1
    [1] => cat*c*2
    [2] => cat*c*1
    [3] => cat*c*2
    [4] => cat*c*3
    [5] => cat*c*1
)

Array
(
    [0] => cat*p*1
    [1] => cat*c*2
    [2] => cat*c*1
    [3] => cat*c*2
    [4] => cat*c*3
    [5] => cat*c*1
)


Array
(
    [0] => cat*p*1
    [1] => cat*c*2
    [2] => cat*c*1
    [3] => cat*c*2
    [4] => cat*c*3
    [5] => cat*c*1
)


Array
(
    [0] => cat*p*1
)

Array
(
    [0] => cat*p*1
    [1] => cat*p*1
    [2] => cat*p*1
    [3] => cat*p*1
    [4] => cat*c*2
    [5] => cat*c*1
    [6] => cat*c*2
    [7] => cat*c*3
    [8] => cat*c*1
)


Array
(
    [0] => cat*p*1
    [1] => cat*p*1
    [2] => cat*p*1
    [3] => cat*p*1
    [4] => cat*c*2
    [5] => cat*c*1
    [6] => cat*c*2
    [7] => cat*c*3
    [8] => cat*c*1
)

Array
(
    [0] => cat*p*1
    [1] => cat*p*1
    [2] => cat*p*1
    [3] => cat*p*1
    [4] => cat*p*1
    [5] => cat*c*2
    [6] => cat*c*1
    [7] => cat*c*2
    [8] => cat*c*3
    [9] => cat*c*1
)


Array
(
    [0] => cat*p*1
    [1] => cat*p*1
    [2] => cat*p*1
    [3] => cat*p*1
    [4] => cat*p*1
    [5] => cat*c*2
    [6] => cat*c*1
    [7] => cat*c*2
    [8] => cat*c*3
    [9] => cat*c*1
)

Array
(
    [0] => cat*p*1
    [1] => cat*p*1
    [2] => cat*p*1
    [3] => cat*p*1
    [4] => cat*p*1
    [5] => cat*p*1
    [6] => cat*c*2
    [7] => cat*c*1
    [8] => cat*c*2
    [9] => cat*c*3
    [10] => cat*c*1
)


Array
(
    [0] => cat*p*1
    [1] => cat*p*1
    [2] => cat*p*1
    [3] => cat*p*1
    [4] => cat*p*1
    [5] => cat*p*1
    [6] => cat*c*2
    [7] => cat*c*1
    [8] => cat*c*2
    [9] => cat*c*3
    [10] => cat*c*1
)


Array
(
    [0] => cat*p*1
    [1] => cat*p*1
    [2] => cat*p*1
    [3] => cat*p*1
    [4] => cat*p*1
    [5] => cat*p*1
    [6] => cat*c*2
    [7] => cat*c*1
    [8] => cat*c*2
    [9] => cat*c*3
    [10] => cat*c*1
)


Array
(
    [0] => cat*p*1
    [1] => cat*p*1
    [2] => cat*p*1
    [3] => cat*p*1
    [4] => cat*p*1
    [5] => cat*p*1
    [6] => cat*c*2
    [7] => cat*c*1
    [8] => cat*c*2
    [9] => cat*c*3
    [10] => cat*c*1
)


Array
(
    [0] => cat*p*1
    [1] => cat*p*1
    [2] => cat*p*1
    [3] => cat*p*1
    [4] => cat*p*1
    [5] => cat*p*1
    [6] => cat*c*2
    [7] => cat*c*1
    [8] => cat*c*2
    [9] => cat*c*3
    [10] => cat*c*1
)
{code}

Here is the same output with the changes applied:
{code}
root@c119d6a06fcb:/var/www/m2clean$  bin/magento cron:run --group=index --bootstrap=standaloneProcessStarted=1
Array
(
)


Array
(
)


Array
(
    [0] => cat*p*1
    [1] => cat*c*2
    [2] => cat*c*1
    [4] => cat*c*3
)

Array
(
    [0] => cat*p*1
    [1] => cat*c*2
    [2] => cat*c*1
    [4] => cat*c*3
)


Array
(
    [0] => cat*p*1
    [1] => cat*c*2
    [2] => cat*c*1
    [4] => cat*c*3
)


Array
(
    [0] => cat*p*1
)

Array
(
    [0] => cat*p*1
    [4] => cat*c*2
    [5] => cat*c*1
    [7] => cat*c*3
)


Array
(
    [0] => cat*p*1
    [4] => cat*c*2
    [5] => cat*c*1
    [7] => cat*c*3
)

Array
(
    [0] => cat*p*1
    [5] => cat*c*2
    [6] => cat*c*1
    [8] => cat*c*3
)


Array
(
    [0] => cat*p*1
    [5] => cat*c*2
    [6] => cat*c*1
    [8] => cat*c*3
)

Array
(
    [0] => cat*p*1
    [6] => cat*c*2
    [7] => cat*c*1
    [9] => cat*c*3
)

Array
(
    [0] => cat*p*1
    [6] => cat*c*2
    [7] => cat*c*1
    [9] => cat*c*3
)

Array
(
    [0] => cat*p*1
    [6] => cat*c*2
    [7] => cat*c*1
    [9] => cat*c*3
)

Array
(
    [0] => cat*p*1
    [6] => cat*c*2
    [7] => cat*c*1
    [9] => cat*c*3
)


Array
(
    [0] => cat*p*1
    [6] => cat*c*2
    [7] => cat*c*1
    [9> => cat*c*3
)

```

### Steps to reproduce (*)
1. Go to **System* - *Index Management**
1. **Select All* *Indexers_  and change the _Actions* to *Update by Schedule**
1. Go to **Catalog* - *Categories**, and change the *Category Name* for at least two Categories, for example
1. Go to **Catalog* - *Products** and change the *Price, Quantity, Weight* for at least two Products
1. Run `bin/magento cron:run --group=index --bootstrap=standaloneProcessStarted=1`
1. Observe the return of `\Magento\Framework\Indexer\CacheContext::getIdentities`

### Expected result (*)
before

### Actual Result (*)
after

### PR DESCRIPTION
**Pull request by siimm**
### Description (*)
If a product is saved in admin - changed name, price, stock, weight, added category to product. Then after indexing, the same tags are flushed several times.
Here is an example of default Magento behaviour on cleaning cache tags during indexing (different arrays belong to different indexers cleaning the cache):
```
root@c119d6a06fcb:/var/www/m2clean$  bin/magento cron:run --group=index --bootstrap=standaloneProcessStarted=1

Array
(
)


Array
(
)


Array
(
    [0] => cat_p_1
    [1] => cat_c_2
    [2] => cat_c_1
    [3] => cat_c_2
    [4] => cat_c_3
    [5] => cat_c_1
)

Array
(
    [0] => cat_p_1
    [1] => cat_c_2
    [2] => cat_c_1
    [3] => cat_c_2
    [4] => cat_c_3
    [5] => cat_c_1
)


Array
(
    [0] => cat_p_1
    [1] => cat_c_2
    [2] => cat_c_1
    [3] => cat_c_2
    [4] => cat_c_3
    [5] => cat_c_1
)


Array
(
    [0] => cat_p_1
)

Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_c_2
    [5] => cat_c_1
    [6] => cat_c_2
    [7] => cat_c_3
    [8] => cat_c_1
)


Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_c_2
    [5] => cat_c_1
    [6] => cat_c_2
    [7] => cat_c_3
    [8] => cat_c_1
)

Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_p_1
    [5] => cat_c_2
    [6] => cat_c_1
    [7] => cat_c_2
    [8] => cat_c_3
    [9] => cat_c_1
)


Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_p_1
    [5] => cat_c_2
    [6] => cat_c_1
    [7] => cat_c_2
    [8] => cat_c_3
    [9] => cat_c_1
)

Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_p_1
    [5] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [8] => cat_c_2
    [9] => cat_c_3
    [10] => cat_c_1
)


Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_p_1
    [5] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [8] => cat_c_2
    [9] => cat_c_3
    [10] => cat_c_1
)


Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_p_1
    [5] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [8] => cat_c_2
    [9] => cat_c_3
    [10] => cat_c_1
)


Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_p_1
    [5] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [8] => cat_c_2
    [9] => cat_c_3
    [10] => cat_c_1
)


Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_p_1
    [5] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [8] => cat_c_2
    [9] => cat_c_3
    [10] => cat_c_1
)
```

Here is the same output with the changes applied:
```
root@c119d6a06fcb:/var/www/m2clean$  bin/magento cron:run --group=index --bootstrap=standaloneProcessStarted=1
Array
(
)


Array
(
)


Array
(
    [0] => cat_p_1
    [1] => cat_c_2
    [2] => cat_c_1
    [4] => cat_c_3
)

Array
(
    [0] => cat_p_1
    [1] => cat_c_2
    [2] => cat_c_1
    [4] => cat_c_3
)


Array
(
    [0] => cat_p_1
    [1] => cat_c_2
    [2] => cat_c_1
    [4] => cat_c_3
)


Array
(
    [0] => cat_p_1
)

Array
(
    [0] => cat_p_1
    [4] => cat_c_2
    [5] => cat_c_1
    [7] => cat_c_3
)


Array
(
    [0] => cat_p_1
    [4] => cat_c_2
    [5] => cat_c_1
    [7] => cat_c_3
)

Array
(
    [0] => cat_p_1
    [5] => cat_c_2
    [6] => cat_c_1
    [8] => cat_c_3
)


Array
(
    [0] => cat_p_1
    [5] => cat_c_2
    [6] => cat_c_1
    [8] => cat_c_3
)

Array
(
    [0] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [9] => cat_c_3
)

Array
(
    [0] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [9] => cat_c_3
)

Array
(
    [0] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [9] => cat_c_3
)

Array
(
    [0] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [9] => cat_c_3
)


Array
(
    [0] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [9] => cat_c_3
)

```

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. Observe the output of \Magento\Framework\Indexer\CacheContext::getIdentities during indexer run, same tags are flushed up to 10 times per one indexer instance
2. ...

### Questions or comments
Together with https://github.com/magento/magento2/pull/29699 it will probably reduce the amount of cache flushing significatly

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#29890: Prevent flushing same tags several times

### CODE REVIEW

**Comment by sidolov** [comment]
@magento create issue

**Comment by rogyar** [comment]
Hi @siimm. Thank you for your collaboration. Nice catch! 
Could I ask you to cover your change with a simple unit test, please? The unit test will assert that the method returns only unique values. 

Thank you!

**Comment by siimm** [comment]
Hi @rogyar, attached a minimalistic test case to this PR
https://github.com/magento/magento2/pull/29885/commits/f17148f902309ddfd6868b17b7183099222084d8

**Comment by lenaorobei** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/Indexer/Test/Unit/CacheContextTest.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                ['cat_p_1']
+            ],
+            'same entities with overlapping ids' => [
+                [['cat_p' => [1, 2, 3]], ['cat_p' => [3]]],
+                ['cat_p_1', 'cat_p_2', 'cat_p_3']
+            ]
+        ];
+    }
+}
+
```
Comment (BarnyShergold) [inline lib/internal/Magento/Framework/Indexer/Test/Unit/CacheContextTest.php]: Extra Blank line is causing test fails

File: lib/internal/Magento/Framework/Indexer/CacheContext.php [hunk 2]
```diff
@@ -73,6 +73,6 @@ public function getIdentities()
                 $identities[] = $cacheTag . '_' . $id;
             }
         }
-        return array_merge($identities, array_unique($this->tags));
+        return array_unique(array_merge($identities, array_unique($this->tags));
```
Comment (BarnyShergold) [inline lib/internal/Magento/Framework/Indexer/CacheContext.php]: Missing a closing parenthesis, should be ))); 
Cause of most of the test failures

**Review by BarnyShergold — CHANGES_REQUESTED** [review]

**Comment by gabrieldagama** [comment]
The risk was set to`medium` due to: the suggested changes may affect other areas/modules.

**Comment by lenaorobei** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by rogyar — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @rogyar, thank you for the review.
[ENGCOM-8237](https://jira.corp.magento.com/browse/ENGCOM-8237) has been created to process this Pull Request

**Review by BarnyShergold — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @BarnyShergold, thank you for the review.
[ENGCOM-8237](https://jira.corp.magento.com/browse/ENGCOM-8237) has been created to process this Pull Request

**Comment by engcom-Bravo** [comment]
:x: QA Failed

Hello @siimm 
Looks like the changes provided in this PR do not solve the original issue. I'm getting the same Output before and after switching to the PR
cat1

Please, see the Manual treating scenario 
1. Go to **System** - **Index Management**
2. **Select All** _Indexers_  and change the _Actions_ to **Update by Schedule**
3. Go to **Catalog** - **Categories**, and change the _Category Name_ for at least two Categories, for example
4. Go to **Catalog** - **Products** and change the _Price, Quantity, Weight_ for at least two Products
5. Run `bin/magento cron:run --group=index --bootstrap=standaloneProcessStarted=1`
6. Observe the output of `\Magento\Framework\Indexer\CacheContext::getIdentities`

**Review by engcom-Bravo — CHANGES_REQUESTED** [review]
:x: QA Failed

Hello @siimm 
Looks like the changes provided in this PR do not solve the original issue. I'm getting the same Output before and after switching to the PR
cat1

Please, see the Manual testing scenario 
1. Go to **System** - **Index Management**
2. **Select All** _Indexers_  and change the _Actions_ to **Update by Schedule**
3. Go to **Catalog** - **Categories**, and change the _Category Name_ for at least two Categories, for example
4. Go to **Catalog** - **Products** and change the _Price, Quantity, Weight_ for at least two Products
5. Run `bin/magento cron:run --group=index --bootstrap=standaloneProcessStarted=1`
6. Observe the output of `\Magento\Framework\Indexer\CacheContext::getIdentities`

**Comment by siimm** [comment]
@engcom-Bravo 
Did you check the values what the **getIdentities** method returns? - from the screenshot I can assume the values were inspected inside of the method itself - the $ids variable is hinting that.

**Comment by engcom-Bravo** [comment]
:heavy_check_mark: QA Passed

Further investigation discovered that the changes provided in this PR, fix the original issue
### Manual Testing scenario
1. Go to **System** - **Index Management**
2. **Select All** _Indexers_  and change the _Actions_ to **Update by Schedule**
3. Go to **Catalog** - **Categories**, and change the _Category Name_ for at least two Categories, for example
4. Go to **Catalog** - **Products** and change the _Price, Quantity, Weight_ for at least two Products
5. Run `bin/magento cron:run --group=index --bootstrap=standaloneProcessStarted=1`
6. Observe the return of `\Magento\Framework\Indexer\CacheContext::getIdentities`

**BEFORE** applying changes
before

**AFTER** switching to the PR
after

**Comment by engcom-Bravo** [comment]
@rogyar @BarnyShergold Can you please review the PR again?

File: lib/internal/Magento/Framework/Indexer/Test/Unit/CacheContextTest.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        $this->object = new CacheContext();
+    }
+
+    /**
+     * @param array $tagsData
+     * @param array $expected
+     * @dataProvider getTagsDataProvider
+     */
+    public function testUniqueTags($tagsData, $expected)
```
Comment (rogyar) [inline lib/internal/Magento/Framework/Indexer/Test/Unit/CacheContextTest.php:31]: Just a recommendation for the future to use type hinting whenever it's possible. It's a very useful habit in PHP 7.x

**Review by rogyar — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @rogyar, thank you for the review.
[ENGCOM-8237](https://jira.corp.magento.com/browse/ENGCOM-8237) has been created to process this Pull Request

**Pull request merged**